# Data preparation

In [ ]:
# Get the data
from math  
import re
import pandas as pd

df = pd.read_csv('data_set/1429_1.csv',low_memory=False)
df.head(5)

id                                               name  \
0  AVqkIhwDv8e3D1O-lebb  All-New Fire HD 8 Tablet, 8 HD Display, Wi-Fi,...   
1  AVqkIhwDv8e3D1O-lebb  All-New Fire HD 8 Tablet, 8 HD Display, Wi-Fi,...   
2  AVqkIhwDv8e3D1O-lebb  All-New Fire HD 8 Tablet, 8 HD Display, Wi-Fi,...   
3  AVqkIhwDv8e3D1O-lebb  All-New Fire HD 8 Tablet, 8 HD Display, Wi-Fi,...   
4  AVqkIhwDv8e3D1O-lebb  All-New Fire HD 8 Tablet, 8 HD Display, Wi-Fi,...   

        asins   brand                                         categories  \
0  B01AHB9CN2  Amazon  Electronics,iPad & Tablets,All Tablets,Fire Ta...   
1  B01AHB9CN2  Amazon  Electronics,iPad & Tablets,All Tablets,Fire Ta...   
2  B01AHB9CN2  Amazon  Electronics,iPad & Tablets,All Tablets,Fire Ta...   
3  B01AHB9CN2  Amazon  Electronics,iPad & Tablets,All Tablets,Fire Ta...   
4  B01AHB9CN2  Amazon  Electronics,iPad & Tablets,All Tablets,Fire Ta...   

                                                keys manufacturer  \
0  841667104676,amazon/53004484,amazon/b01ahb9cn2...       Amazon   
1  841667104676,amazon/53004484,amazon/b01ahb9cn2...       Amazon   
2  841667104676,amazon/53004484,amazon/b01ahb9cn2...       Amazon   
3  841667104676,amazon/53004484,amazon/b01ahb9cn2...       Amazon   
4  841667104676,amazon/53004484,amazon/b01ahb9cn2...       Amazon   

               reviews.date     reviews.dateAdded  \
0  2017-01-13T00:00:00.000Z  2017-07-03T23:33:15Z   
1  2017-01-13T00:00:00.000Z  2017-07-03T23:33:15Z   
2  2017-01-13T00:00:00.000Z  2017-07-03T23:33:15Z   
3  2017-01-13T00:00:00.000Z  2017-07-03T23:33:15Z   
4  2017-01-12T00:00:00.000Z  2017-07-03T23:33:15Z   

                                    reviews.dateSeen  ... reviews.doRecommend  \
0  2017-06-07T09:04:00.000Z,2017-04-30T00:45:00.000Z  ...                True   
1  2017-06-07T09:04:00.000Z,2017-04-30T00:45:00.000Z  ...                True   
2  2017-06-07T09:04:00.000Z,2017-04-30T00:45:00.000Z  ...                True   
3  2017-06-07T09:04:00.000Z,2017-04-30T00:45:00.000Z  ...                True   
4  2017-06-07T09:04:00.000Z,2017-04-30T00:45:00.000Z  ...                True   

  reviews.id  reviews.numHelpful  reviews.rating  \
0        NaN                 0.0             5.0   
1        NaN                 0.0             5.0   
2        NaN                 0.0             5.0   
3        NaN                 0.0             4.0   
4        NaN                 0.0             5.0   

                                  reviews.sourceURLs  \
0  http://reviews.bestbuy.com/3545/5620406/review...   
1  http://reviews.bestbuy.com/3545/5620406/review...   
2  http://reviews.bestbuy.com/3545/5620406/review...   
3  http://reviews.bestbuy.com/3545/5620406/review...   
4  http://reviews.bestbuy.com/3545/5620406/review...   

                                        reviews.text  \
0  This product so far has not disappointed. My c...   
1  great for beginner or experienced person. Boug...   
2  Inexpensive tablet for him to use and learn on...   
3  I've had my Fire HD 8 two weeks now and I love...   
4  I bought this for my grand daughter when she c...   

                             reviews.title reviews.userCity  \
0                                   Kindle              NaN   
1                                very fast              NaN   
2  Beginner tablet for our 9 year old son.              NaN   
3                                  Good!!!              NaN   
4                Fantastic Tablet for kids              NaN   

   reviews.userProvince  reviews.username  
0                   NaN           Adapter  
1                   NaN            truman  
2                   NaN             DaveZ  
3                   NaN            Shacks  
4                   NaN         explore42  

[5 rows x 21 columns]

## Preprocess data : Remove Nan, html tags and extra whitespaces

In [ ]:

# Choose input features from the data
df_input = df.loc[:, ['name', 'categories', 'reviews.rating', 'reviews.text']]

print(df_input.head())

# preprocess the input features
df_input['name'] = df_input['name'].str.strip()
df_input['categories'] = df_input['categories'].str.lower().str.strip()
df_input['reviews.text'] = df_input['reviews.text'].str.lower().str.strip()
df_input['reviews.rating'] = df_input['reviews.rating'].astype(str)



# Remove rows with NaN values in the specified columns
df_input = df_input.dropna(subset=['name', 'categories', 'reviews.rating', 'reviews.text'])


 # Remove html tags from reviews
def remove_html_tags(text):
    clean = re.compile('<.*?>')
    return re.sub(clean, '', text)

df_input['reviews.text'] = df_input['reviews.text'].apply(remove_html_tags)


# Remove extra whitespace from reviews
def remove_extra_whitespace(text):
    return re.sub(r'\s+', ' ', text).strip()

df_input['reviews.text'] = df_input['reviews.text'].apply(remove_extra_whitespace)

# Remove leading and trailing whitespace from product names and categories
df_input['name'] = df_input['name'].str.strip()
df_input['categories'] = df_input['categories'].str.strip() 

# review columns
print(df_input[['reviews.text', 'reviews.rating']].head(3))

In [5]:

# Remove extra whitespace from reviews
def remove_extra_whitespace(text):
    return re.sub(r'\s+', ' ', text).strip()

df_input['reviews.text'] = df_input['reviews.text'].apply(remove_extra_whitespace)

# Remove leading and trailing whitespace from product names and categories
df_input['name'] = df_input['name'].str.strip()
df_input['categories'] = df_input['categories'].str.strip() 

# review columns
print(df_input[['reviews.text', 'reviews.rating']].head(3))

                                        reviews.text reviews.rating
0  this product so far has not disappointed. my c...            5.0
1  great for beginner or experienced person. boug...            5.0
2  inexpensive tablet for him to use and learn on...            5.0


In [6]:
# Create a new column for the training data
df_input['prompt'] = 'Product: ' + df_input['name'] + '\nCategory: ' + df_input['categories'] + '\nRating: ' + df_input['reviews.rating'] 

print('\n',df_input['prompt'][1])


 Product: All-New Fire HD 8 Tablet, 8 HD Display, Wi-Fi, 16 GB - Includes Special Offers, Magenta
Category: electronics,ipad & tablets,all tablets,fire tablets,tablets,computers & tablets
Rating: 5.0


## Combine training and target into one column

In [ ]:

# split data into training and target and combinate them in one column, a feature in GPT-2 that we can use in comibine the prompt and the reviews text
df_input['training_data'] = df_input['prompt'] + ' ' + df_input['reviews.text']

# Load the training data into JSON to be used later in Hugging Face
import json

with open("training_data.jsonl", "w") as f:
    for _, row in df_input.iterrows():
        json.dump({"text": row['training_data']}, f)
        f.write("\n")

from datasets import load_dataset

dataset = load_dataset('json', data_files='training_data.jsonl', split='train')

print('The dataset (review + context) is now a dictionary in JSON format:', dataset[0]) 

# Split data into training and test sets


In [8]:
# Split data into training and test sets

split_dataset = dataset.train_test_split(test_size=0.2)

train_dataset = split_dataset['train']
test_dataset = split_dataset['test']

print("Train dataset size:", len(train_dataset))
print("Test dataset size:", len(test_dataset))


Train dataset size: 22319
Test dataset size: 5580


# Train the model

# tokenize the dataset


In [9]:
from transformers import GPT2Tokenizer

# Load GPT-2 tokenizer
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
tokenizer.pad_token = tokenizer.eos_token  # Important for GPT-2, otherwise an error will occur

# Tokenization function
def tokenize_function(example):
    return tokenizer(
        example["text"],
        truncation=True,
        padding="max_length",
        max_length=512
    )

# Tokenize datasets
tokenized_train = train_dataset.map(tokenize_function, batched=True)
tokenized_test = test_dataset.map(tokenize_function, batched=True)
tokenized_trainttt = tokenize_function(train_dataset[0])


Map: 100%|██████████| 5580/5580 [00:08<00:00, 655.39 examples/s]


In [ ]:
print(train_dataset[0], '\n')


print("Tokenized train dataset size:", tokenized_train['input_ids'][0], '\n')

print(tokenizer.decode(tokenized_train["input_ids"][0]))  # Back to text


## Import and load the model for training


In [11]:
from transformers import GPT2LMHeadModel

model = GPT2LMHeadModel.from_pretrained("gpt2")
model.resize_token_embeddings(len(tokenizer))  # Adjust for pad token


Embedding(50257, 768)

In [12]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer, mlm=False, mlm_probability=0.15
)

import transformers
print(transformers.__version__)



4.55.0


## Setting training argument + training the baseline gpt2

### The model is stopped on 24min 21 sec. dur to the large amount of time in training

In [13]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="./gpt2-review-gen",
    eval_strategy="epoch",
    learning_rate=5e-5,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    num_train_epochs=3,
    weight_decay=0.01,
    save_total_limit=2,
    logging_dir="./logs",
    logging_steps=100,
    push_to_hub=False
)


from transformers import Trainer

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train,
    eval_dataset=tokenized_test,
    tokenizer=tokenizer,
    data_collator=data_collator
)


trainer.train()


C:\Users\User\AppData\Local\Temp\ipykernel_324\61388656.py:20: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(
C:\Users\User\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\torch\utils\data\dataloader.py:666: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  warnings.warn(warn_msg)
`loss_type=None` was set in the config but it is unrecognised.Using the default loss: `ForCausalLMLoss`.


KeyboardInterrupt: 

In [14]:
trainer.save_model("./gpt2-review-gen")
tokenizer.save_pretrained("./gpt2-review-gen")



('./gpt2-review-gen\\tokenizer_config.json',
 './gpt2-review-gen\\special_tokens_map.json',
 './gpt2-review-gen\\vocab.json',
 './gpt2-review-gen\\merges.txt',
 './gpt2-review-gen\\added_tokens.json')

# Evaluation

## Determine the perplexity for the base-line gpt2 model

In [19]:
small_eval = tokenized_test.shuffle(seed=42).select(range(100))
eval_results = trainer.evaluate(eval_dataset=small_eval)
perplexity = math.exp(eval_results["eval_loss"])
print(f"Perplexity: {perplexity:.2f}")


C:\Users\User\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\torch\utils\data\dataloader.py:666: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  warnings.warn(warn_msg)


Perplexity: 2.93


## Now, the perlexity is 2.93. The goal is to decrease it

### We decide to put in evidence the Context and Review into a new column with a clear format of Context and Review: A fine-tuned gpt2 model

In [41]:
df_input.head(3)


def format_training_example(column):
    return (
        f"CONTEXT:\n"
        f"Product: {column['name']}\n"
        f"Category: {column['categories']}\n"
        f"Rating: {column['reviews.rating']}\n\n"
        f"REVIEW:\n"
        f"{column['reviews.text']}"
    )

df_input['Context, Review'] = df_input.apply(format_training_example, axis=1)
df_input.head(3)
df_input['Context, Review'][5]


with open("Context, Review.jsonl", "w") as f:
    for _, row in df_input.iterrows():
        json.dump({"text": row['Context, Review']}, f)
        f.write("\n")

from datasets import load_dataset

dataset = load_dataset('json', data_files='Context, Review.jsonl', split='train')

print(dataset[0]) 


# Split data into training and test sets

split_dataset = dataset.train_test_split(test_size=0.2)

train_dataset_context_Review = split_dataset['train']
test_dataset_context_Review = split_dataset['test']


# Tokenize datasets
tokenized_train = train_dataset_context_Review.map(tokenize_function, batched=True)
tokenized_test = test_dataset_context_Review.map(tokenize_function, batched=True)
tokenized_trainttt = tokenize_function(train_dataset_context_Review[0])


Generating train split: 27899 examples [00:00, 468684.34 examples/s]


{'text': 'CONTEXT:\nProduct: All-New Fire HD 8 Tablet, 8 HD Display, Wi-Fi, 16 GB - Includes Special Offers, Magenta\nCategory: electronics,ipad & tablets,all tablets,fire tablets,tablets,computers & tablets\nRating: 5.0\n\nREVIEW:\nthis product so far has not disappointed. my children love to use it and i like the ability to monitor control what content they see with ease.'}


Map: 100%|██████████| 5580/5580 [00:11<00:00, 494.47 examples/s]


## Setting training arguments + training: We stop training on 24mins.21 sec, the same training time for the previous simulation

In [46]:

training_args = TrainingArguments(
    output_dir="./results/run_1",
    eval_strategy="epoch",
    learning_rate=5e-5,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    num_train_epochs=3,
    weight_decay=0.01,
    save_total_limit=2,
    logging_dir="./logs",
    logging_steps=100,
    push_to_hub=False
)


trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train,
    eval_dataset=tokenized_test,
    tokenizer=tokenizer,
    data_collator=data_collator
)


trainer.train()



C:\Users\User\AppData\Local\Temp\ipykernel_324\2153007247.py:16: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(
C:\Users\User\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\torch\utils\data\dataloader.py:666: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  warnings.warn(warn_msg)


KeyboardInterrupt: 

In [48]:
trainer.save_model("./results/run_1")
tokenizer.save_pretrained("./results/run_1")


('./results/run_1\\tokenizer_config.json',
 './results/run_1\\special_tokens_map.json',
 './results/run_1\\vocab.json',
 './results/run_1\\merges.txt',
 './results/run_1\\added_tokens.json')

## Determine perplexity for fine-tuned Gpt2

In [49]:
small_eval = tokenized_test.shuffle(seed=42).select(range(100))
eval_results = trainer.evaluate(eval_dataset=small_eval)
perplexity = math.exp(eval_results["eval_loss"])
print(f"Perplexity: {perplexity:.2f}")


Perplexity: 2.48


## Examples generation: ***before vs after***

In [ ]:
from transformers import pipeline

generator_baseline = pipeline("text-generation", model="./gpt2-review-gen")

generator_finetuned = pipeline("text-generation", model="./results/run_1", tokenizer=tokenizer)


prompt = "Product: Logitech Wireless Mouse\nCategory: electronics\nRating: 5\nReview:"

print(generator_baseline(prompt, max_length=100, num_return_sequences=1)[0]['generated_text'])

print(generator_finetuned(prompt, max_length=100, num_return_sequences=1)[0]['generated_text'])


Device set to use cpu
Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=100) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=256) and `max_length`(=100) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Product: Logitech Wireless Mouse
Category: electronics
Rating: 5
Review: 5.0 this is an excellent purchase for a $300 price. it is a bit of a hassle to get in and out. the wireless charging is good but the wireless is not as fast as i expected. the wireless is not as fast as i expected. i have been using it for about 6 months now and i am confident that i will continue using it in future. my only complaint is the battery. i am happy with the purchase price. the wireless is good but not fast enough. i am pleased with the purchase price. i am happy with the purchase price. the wireless is good but not fast enough. i am pleased with the purchase price. the wireless is good but not fast enough. i am pleased with the purchase price. the wireless is good but not fast enough. i am pleased with the purchase price. the wireless is good but not fast enough. i am pleased with the purchase price. the wireless is good but not fast enough. i am happy with the purchase price.


Rating: 5.0 this is de